packages/kbn-subscription-tracking/src/use_impression.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { useEffect } from 'react';
+import { isValidContext } from './helpers';
+import { useServices } from './services';
+import { EVENT_NAMES, SubscriptionContextData } from '../types';
+
+/**
+ * Sends an impression event with the given context.
+ *
+ * Note: impression events are throttled and will not fire more
+ * often than once every 30 seconds.
+ */
+export const useImpression = (context: SubscriptionContextData) => {
+  const { analyticsClient } = useServices();
+
+  useEffect(() => {
+    if (!isValidContext(context)) {
+      // eslint-disable-next-line no-console
+      console.error('The provided subscription context is invalid', context);
+      return;
+    }
+    if (!isCoolingDown(context)) {
+      analyticsClient.reportEvent(EVENT_NAMES.IMPRESSION, context);
+      coolDown(context);
+    }
+  }, [analyticsClient, context]);
+};
+
+/**
+ * Impressions from the same context should not fire more than once every 30 seconds.
+ * This prevents logging too many impressions in case a page is reloaded often or
+ * if the user is navigating back and forth rapidly.
+ */
+export const coolDownTimeMs = 30 * 1000;
+let impressionCooldown = new WeakMap<SubscriptionContextData, number>();
+
+function isCoolingDown(context: SubscriptionContextData) {
+  const previousLog = impressionCooldown.get(context);
+
+  // we logged before and we are in the cooldown period
+  return previousLog && Date.now() - previousLog < coolDownTimeMs;
+}
+
+function coolDown(context: SubscriptionContextData) {
+  impressionCooldown.set(context, Date.now());
+}
+
+export function resetCoolDown() {
```

Comment:
As far as I see this function is only exported to be used in tests. It's better to avoid it. This way the logic here can be simplified

```ts
import * as _ from 'lodash';

...

const reportAnalyticsThrottled = useMemo(() => _.throttle(() => analyticsClient.reportEvent(EVENT_NAMES.IMPRESSION, context), 30 * 1000), [analyticsClient]);

...

useEffect(() => {
  ...

  reportAnalyticsThrottled();

  return reportAnalyticsThrottled.cancel;
}, [analyticsClient, context, reportAnalyticsThrottled]);
```

Reply:
@maximpn I am not sure this approach will work as it will cancel the throttle when the component is unmounted. With the current approach, the cooldown of 30 seconds is consistent even when the component unmounts.

However, I agree that currently it looks weird to have this function only used in tests. It might come in handy though for some edge cases later down the line when the cooldown should be cancelled earlier for example.

Reply:
If it's not necessary to cancel throttling it's enough to remove `return reportAnalyticsThrottled.cancel;` line. Though the logic canceling the timer upon unmounting is quite straightforward as the time limit hasn't reached then logging isn't expected. It will log on the next mount if it's remounting. Otherwise if remounting happens too often it's rather a problem in the driver code which has to be fixed.

> However, I agree that currently it looks weird to have this function only used in tests. It might come in handy though for some edge cases later down the line when the cooldown should be cancelled earlier for example.

It's quite hard to shrink an API than extending it. If something isn't needed the best way is to avoid exposing it. Whenever there is a request to reset the cooldown timer the functionality can be added.

Does it sound reasonable? The main point I pursue simpler is better. Having less code reduces maintainability costs by simplifying its reading and reducing number of tests.